static/js/message_list_view.js

```diff
@@ -1445,4 +1445,87 @@ export class MessageListView {
             message_container.status_message = false;
         }
     }
+
+    update_sticky_recipient_headers() {
+        const rows_length = this._rows.size;
+        if (!rows_length) {
+            /* No headers are present */
+            return;
+        }
+        /* Intentionally remove sticky headers class here to make calculations simpler. */
+        $(".sticky_header").removeClass("sticky_header");
+        /* visible_top is navbar top position + height for us. */
+        const visible_top = $message_viewport.message_viewport_info().visible_top;
+        /* We need date to be properly visible on the header, so partially visible headers
+           who are about to be scrolled out of view are not acceptable. */
+        const header_about_to_be_partially_hidden = visible_top - 1;
+
+        function is_sticky(header) {
+            const header_props = header.getBoundingClientRect();
+            const margin_between_recipient_rows = 10;
+            const header_is_sticky_or_about_to_be_sticky =
+                visible_top + header_props.height + margin_between_recipient_rows;
+            /* If two headers next to each other are completely visible
+               (message header at top has no visible content), we don't mind showing
+               date on any of them. Which header is chosen will depend on which
+               comes first when iterating on the headers. */
+            return (
+                header_about_to_be_partially_hidden <= header_props.top &&
+                header_props.top <= header_is_sticky_or_about_to_be_sticky
+            );
+        }
+
+        const $table = rows.get_table(this.table_name);
+        const $headers = $table.find(".message_header");
+        const iterable_headers = $headers.toArray();
+        /* To get to the sticky header ASAP, we take help from how far has user scrolled. */
+        if ($message_viewport.scrollTop() > $message_viewport.height() / 2) {
+            iterable_headers.reverse();
+        }
+        let sticky_header;
+        for (const header of iterable_headers) {
+            if (is_sticky(header)) {
+                sticky_header = header;
+                break;
+            }
+        }
+
+        /* Set correct date for the sticky_header. */
+        let $message_row;
+        if (!sticky_header) {
+            /* If user at the top of the scroll container, there is no sticky
+               header, so we just set date of the header to the first visible message. */
+            sticky_header = $headers[0];
+            $message_row = $($headers[0].nextElementSibling);
```

Comment:
[`$headers.next()`](https://api.jquery.com/next/#next-selector)?

Reply:
It is either `$($headers[0]).next()` or `$($headers[0].nextElementSibling);` of which the second one seems more readable, so I went with it.

Reply:
Ok!